lib/ansible/modules/windows/win_get_url.ps1

```diff
@@ -155,28 +332,82 @@ Function Download-File($module, $url, $dest, $headers, $credentials, $timeout, $
+
+        # Checksum verification for downloaded file
+        if ($checksum) {
+            # Check both hashes are the same
+            if ($tmpDestHash -ne $checksum) {
+                throw [string]("The checksum for {0} did not match '{1}', it was '{2}'" -f $dest, $checksum, $tmpDestHash)
+            } else {
+                if(Test-Path -LiteralPath $dest) {
+                    $destHash = Get-FileChecksum -path $dest -algorithm $checksum_algorithm
+                    if ($destHash -eq $checksum) {
+                        $module.Result.status_code = 200
+                        $module.Result.msg = 'file already exists'
+                        $module.Result.checksum_src = $tmpDestHash
+                        $module.Result.elapsed = ((Get-Date) - $module_start).TotalSeconds
+                        return
+                    }
+                } else {
+                    Copy-Item -Path $tmpDest -Destination $dest -Force -WhatIf:$module.CheckMode | Out-Null
+                    $diff += "+$dest`n"
+                }
+            }
+        } else {
+            $is_modified = Compare-ModifiedFile -module $module -url $url -dest $tmpDest -credentials $credentials -headers $headers `
+                                                -timeout $timeout -use_proxy $use_proxy -proxy $proxy -check_length $true
+            if ($is_modified) {
+                throw "Source and recieved files size mismatch."
+            }
+            Copy-Item -Path $tmpDest -Destination $dest -Force -WhatIf:$module.CheckMode | Out-Null
+            $diff += "+$dest`n"
+        }
+
+    } Catch [System.Net.WebException] {
+        $module.Result.status_code = [int] $_.Exception.Response.StatusCode
+        $module.Result.elapsed = ((Get-Date) - $module_start).TotalSeconds
+        $module.FailJson("Error downloading '$url' to '$dest': $($_.Exception.Message)", $_)
+    } Catch {
+        $module.Result.elapsed = ((Get-Date) - $module_start).TotalSeconds
+        $module.Result.checksum_dest = $destHash
+        $module.FailJson("Unknown error downloading '$url' to '$dest': $($_.Exception.Message)", $_)
+    }
+    Finally {
+        if (Test-Path -LiteralPath $tmpDest) {
+            Remove-Item -Path $tmpDest | Out-Null
+            $diff += "-$tmpDest`n"
+
+            $module.Diff.after.files = $diff
         }
     }
 
     $module.Result.status_code = 200
     $module.Result.changed = $true
     $module.Result.msg = 'OK'
     $module.Result.dest = $dest
+    $module.Result.checksum_src = $tmpDestHash
     $module.Result.elapsed = ((Get-Date) - $module_start).TotalSeconds
+}
 
+# normalise values
+if ($checksum) {
+    $checksum = $checksum.Trim().toLower()
+}
+if ($checksum_algorithm) {
+    $checksum_algorithm = $checksum_algorithm.Trim().toLower()
+}
+if ($checksum_url) {
+    $checksum_url = $checksum_url.Trim().toLower()
 }
 
+$module.Result.checksum_algorithm = $checksum_algorithm
```

Comment:
Don't need to do this. The other values are only there for historical reason but with the new Ansible.Module.AnsibleBasic util this is automatically added under the `invocation` key just like Python modules.

Reply:
Fixed

Reply:
I’d also appreciate it if you could explain how we can get value for checksum_algorithm if it uses default value (when we didn't set `checksum_algorithm` at all)?

Reply:
@jborean93 Could you please take a look?

Reply:
Fixed

Reply:
Sorry @vutkin, it's not something you can necessarily retrieve in the test but because it's just module option invocation vars you should already know what it is.